src/EncodingContext.h

```diff
@@ -2,117 +2,49 @@
-#include "thrift/transport/TBufferTransports.h"
+#include <vector>
 #include "EncodingTypes.h"
-#include "Encoder.h"
-
-class Span;
 
 namespace cppkin
 {
     class Span;
-    class SimpleAnnotation;
+    class Encoder;
 
     class EncoderContext
     {
-    public:
-        virtual ~EncoderContext();
-        virtual std::string ToString() = 0;
-    };
-
-    class EncoderContextThrift : public EncoderContext
-    {
-
-    private:
-        class ContextElement
-        {
-        public:
-            typedef ::Span* pointer;
-            typedef ::Span& reference;
-            typedef apache::thrift::protocol::TBinaryProtocol protocol;
-            typedef apache::thrift::transport::TMemoryBuffer buffer;
-
-            ContextElement(reference span, protocol& _protocol, buffer& _buffer);
-            ~ContextElement();
-            ContextElement(const ContextElement& obj) = delete;
-            ContextElement& operator=(const ContextElement& obj) = delete;
-            std::string ToString() const;
-
-            reference operator*() const;
-            pointer operator->() const;
-        private:
-            std::unique_ptr<::Span> m_span;
-            protocol& m_protocol;
-            buffer& m_buffer;
-        };
 
     public:
-        class EncoderContextThriftIterator
-        {
-        public:
-            typedef EncoderContextThriftIterator self;
-            typedef ContextElement* pointer;
-            typedef ContextElement& reference;
-
-            EncoderContextThriftIterator(const std::list<ContextElement>::iterator& it);
-            ~EncoderContextThriftIterator();
-            reference operator*() const;
-            pointer operator->() const;
-            self& operator++(int);
-            bool operator==(const self& obj) const;
-            bool operator!=(const self& obj) const;
-
-
-        private:
-            std::list<ContextElement>::iterator m_it;
-        };
 
+        class ContextElement {
 
-        class EncoderContextThriftConstIterator
-        {
         public:
-            typedef EncoderContextThriftConstIterator self;
-            typedef const ContextElement* pointer;
-            typedef const ContextElement& reference;
-
-            EncoderContextThriftConstIterator(const std::list<ContextElement>::const_iterator& it);
-            ~EncoderContextThriftConstIterator();
-            reference operator*() const;
-            pointer operator->() const;
-            self& operator++(int);
-            bool operator==(const self& obj) const;
-            bool operator!=(const self& obj) const;
 
+            typedef const Span& reference;
+            ContextElement(Span* span, const Encoder& encoder);
+            std::string ToString() const;
+            reference operator*() const { return *m_span; }
 
         private:
-            std::list<ContextElement>::const_iterator m_it;
+            std::unique_ptr<Span> m_span;
+            const Encoder&        m_encoder;
```

Comment:
Since you are having reference, copy constructor, default constructor and copy assignment needs to be deleted, there is no true copy of the state.

Reply:
I disagree. Since const reference has to be initialized in constructor only, there must be non-default constructor. and as soon as you have it there is no default constructor. Similar with default assignment and copy constructor - they just cannot initialize const reference member therefore no need in explicit deletion of them